www/main/components/addons/bani-controller/components/BaniController.jsx

```diff
@@ -0,0 +1,152 @@
+import { useStoreState, useStoreActions } from 'easy-peasy';
+import React, { useState, useEffect, useRef } from 'react';
+import PropTypes from 'prop-types';
+import { remote } from 'electron';
+import isOnline from 'is-online';
+
+import useSocketListeners from '../hooks/use-socket-listeners';
+import BaniControllerItem from './BaniControllerItem';
+import { Overlay } from '../../../../sttm-ui';
+import QrCode from './QrCode';
+
+import { getBaniControllerItems, generateQrCode, shareSync } from '../utils';
+import ConnectionSwitch from './ConnectionSwitch';
+
+const { tryConnection, onEnd } = shareSync;
+
+const { i18n } = remote.require('./app');
+const analytics = remote.getGlobal('analytics');
+
+const BaniController = ({ onScreenClose }) => {
+  const title = 'Mobile device sync';
+  const canvasRef = useRef(null);
+  const syncRef = useRef(null);
+  // Local State
+  const [codeLabel, setCodeLabel] = useState('');
+  const [isFetchingCode, setFetchingCode] = useState(false);
+  const [isAdminPinVisible, setAdminPinVisibility] = useState(true);
+  const [theFirstRun, setTheFirstRun] = useState(true);
+  // Store State
+  const { isListeners, overlayScreen } = useStoreState(state => state.app);
+  const { setOverlayScreen, setListeners } = useStoreActions(actions => actions.app);
+
+  const { adminPin, code, isConnected } = useStoreState(state => state.baniController);
+  const { setAdminPin, setCode, setConnection } = useStoreActions(
+    actions => actions.baniController,
+  );
+
+  const showSyncError = errorMessage => {
+    setCodeLabel(errorMessage);
+    setCode(null);
+    setAdminPin(null);
+  };
+
+  const remoteSyncInit = async () => {
+    setFetchingCode(true);
+
+    // 1. check onlineValue
+    const onlineValue = await isOnline();
+    if (onlineValue) {
+      const newCode = await tryConnection();
+
+      if (newCode) {
+        const newAdminPin = Math.floor(1000 + Math.random() * 8999);
+
+        setCode(newCode);
+        setAdminPin(newAdminPin);
+
+        generateQrCode(canvasRef.current, newCode);
+
+        setConnection(true);
+        setListeners(true);
+      } else {
+        showSyncError(i18n.t('TOOLBAR.SYNC_CONTROLLER.CODE_ERR'));
+      }
+    } else {
+      showSyncError(i18n.t('TOOLBAR.SYNC_CONTROLLER.INTERNET_ERR'));
+    }
+
+    setFetchingCode(false);
+  };
+
+  const syncToggle = async (forceConnect = false) => {
+    if (isConnected && !forceConnect) {
+      // TODO: Needs to remove this DOM interaction
+      document.body.classList.remove('controller-on');
+      setListeners(false);
+      setConnection(false);
+      onEnd(code);
+      setCode(null);
+      setAdminPin(null);
+      analytics.trackEvent('syncStopped', true);
+    } else {
+      await remoteSyncInit();
+    }
+  };
+
+  const toggleLockScreen = () => {
+    if (overlayScreen !== 'lock-screen') {
+      setOverlayScreen('lock-screen');
+    }
+  };
+
+  useEffect(() => {
```

Comment:
if this is just meant to happen the first time, a useEffect with an empty dep array would be better

Reply:
Aaah, thank you so much. Have been searching for this.